vpr/src/route/route_timing.h

```diff
@@ -10,9 +10,196 @@
+// node (which is returned) through the rr_node_route_inf.  See
+// update_traceback as an example of this tracing.
+class ConnectionRouter {
+  public:
+    ConnectionRouter(
+        const RouterLookahead& router_lookahead,
+        const t_rr_graph_storage& rr_nodes,
+        const std::vector<t_rr_rc_data>& rr_rc_data,
+        const std::vector<t_rr_switch_inf>& rr_switch_inf,
+        std::vector<t_rr_node_route_inf>& rr_node_route_inf)
+        : router_lookahead_(router_lookahead)
+        , rr_nodes_(&rr_nodes)
+        , rr_rc_data_(rr_rc_data.data())
+        , rr_switch_inf_(rr_switch_inf.data())
+        , rr_node_route_inf_(rr_node_route_inf.data()) {}
+
+    // Clear's the modified list.  Should be called after reset_path_costs
+    // have been called.
+    void clear_modified_rr_node_info() {
+        modified_rr_node_inf_.clear();
+    }
+
+    // Reset modified data in rr_node_route_inf based on modified_rr_node_inf.
+    void reset_path_costs() {
+        ::reset_path_costs(modified_rr_node_inf_);
+    }
+
+    // Finds a path from the route tree rooted at rt_root to sink_node
+    //
+    // This is used when you want to allow previous routing of the same net to
+    // serve as valid start locations for the current connection.
+    //
+    // Returns either the last element of the path, or nullptr if no path is
+    // found
+    t_heap* timing_driven_route_connection_from_route_tree(
+        t_rt_node* rt_root,
+        int sink_node,
+        const t_conn_cost_params cost_params,
+        t_bb bounding_box,
+        RouterStats& router_stats);
+
+    // Finds a path from the route tree rooted at rt_root to sink_node for a
+    // high fanout net.
+    //
+    // Unlike timing_driven_route_connection_from_route_tree(), only part of
+    // the route tree which is spatially close to the sink is added to the heap.
+    t_heap* timing_driven_route_connection_from_route_tree_high_fanout(
+        t_rt_node* rt_root,
+        int sink_node,
+        const t_conn_cost_params cost_params,
+        t_bb bounding_box,
+        const SpatialRouteTreeLookup& spatial_rt_lookup,
+        RouterStats& router_stats);
+
+    // Finds a path from the route tree rooted at rt_root to all sinks
+    // available.
+    //
+    // Each element of the returned vector is a reachable sink.
+    //
+    // If cost_params.astar_fac is set to 0, this effectively becomes
+    // Dijkstra's algorithm with a modified exit condition (runs until heap is
+    // empty).  When using cost_params.astar_fac = 0, for efficiency the
+    // RouterLookahead used should be the NoOpLookahead.
+    std::vector<t_heap> timing_driven_find_all_shortest_paths_from_route_tree(
+        t_rt_node* rt_root,
+        const t_conn_cost_params cost_params,
+        t_bb bounding_box,
+        RouterStats& router_stats);
+
+  private:
+    // Mark that inode has been modified, and needs to be reset in
```

Comment:
Mark that the data associated with rr_node "inode" has been ...

Reply:
done